src/Middleware/HttpCache.php

```diff
@@ -0,0 +1,197 @@
+                'Last-Modified',
+                gmdate('D, d M Y H:i:s', $lastModified) . ' GMT'
+            );
+        }
+        if ($cacheValid) {
+            $response = $this->responseFactory->createResponse(304);
+            $response->getBody()->write('Not Modified');
+            return $response;
+        }
+        return $response;
+    }
+
+    /**
+     * Validates if the HTTP cache contains valid content.
+     * If both Last-Modified and ETag are null, returns false.
+     * @param ServerRequestInterface $request
+     * @param int|null $lastModified the calculated Last-Modified value in terms of a UNIX timestamp.
+     * If null, the Last-Modified header will not be validated.
+     * @param string|null $etag the calculated ETag value. If null, the ETag header will not be validated.
+     * @return bool whether the HTTP cache is still valid.
+     */
+    private function validateCache(ServerRequestInterface $request, ?int $lastModified, ?string $etag): bool
+    {
+        if ($request->hasHeader('If-None-Match')) {
+            // HTTP_IF_NONE_MATCH takes precedence over HTTP_IF_MODIFIED_SINCE
+            // http://tools.ietf.org/html/rfc7232#section-3.3
+            return $etag !== null && \in_array($etag, $this->getETags($request), true);
+        }
+
+        if ($request->hasHeader('If-Modified-Since')) {
+            $headers = $request->getHeader('If-Modified-Since');
+            $header = \reset($headers);
+            return $lastModified !== null && @strtotime($header) >= $lastModified;
+        }
+
+        return false;
+    }
+
+    /**
+     * Generates an ETag from the given seed string.
+     * @param string $seed Seed for the ETag
+     * @return string the generated ETag
+     */
+    private function generateEtag(string $seed): string
+    {
+        $etag = '"' . rtrim(base64_encode(sha1($seed, true)), '=') . '"';
+        return $this->weakEtag ? 'W/' . $etag : $etag;
+    }
+
+    /**
+     * Gets the Etags.
+     *
+     * @param ServerRequestInterface $request
+     * @return array The entity tags
+     */
+    private function getETags(ServerRequestInterface $request): array
+    {
+        if ($request->hasHeader('If-None-Match')) {
+            $headers = $request->getHeader('If-None-Match');
+            $header = \str_replace('-gzip', '', \reset($headers));
+            return \preg_split('/[\s,]+/', $header, -1, PREG_SPLIT_NO_EMPTY) ?: [];
+        }
+
+        return [];
+    }
+
+    public function setLastModified(callable $lastModified): void
+    {
+        $this->lastModified = $lastModified;
+    }
+
+    public function setEtagSeed(callable $etagSeed): void
+    {
+        $this->etagSeed = $etagSeed;
+    }
+
+    public function setWeakTag(bool $weakTag): void
+    {
+        $this->weakEtag = $weakTag;
+    }
+
+    public function setParams($params): void
+    {
+        $this->params = $params;
+    }
+
+    public function setCacheControlHeader(?string $header): void
```

Comment:
A builder class would be practical. For example:
```php
..->setCacheControlHeader((string) (new CacheControlHeaderBuilder())
    ->public()
    ->maxAge(300)
    ->sharedMaxAge(60)
  );
```

Reply:
@samdark

Reply:
I think the header builder can be implemented in [yiisoft/http](https://github.com/yiisoft/http)